bindings_ffi/README.md

```diff
@@ -1,41 +1,42 @@
+- Install Docker
+- Install Cross for zero setup cross-platform builds: `cargo install cross --git https://github.com/cross-rs/cross`
+- Run `./cross_build.sh` to cross-compile (this is SLOW)
 
-- Run `make framework`
+`Cross` allows us to run cross-platform builds without needing to download all of the relevant toolchains and should work regardless of your host OS. It is possible that the build time can be improved by building natively without Cross.
 
-## Optional Steps for xmtp-ios integration
+# Uniffi
 
-- Get set up with Xcode
-- Clone [xmtp-ios](https://github.com/xmtp/xmtp-ios)
-- Open Xcode, then open the `xmtp-ios` folder at the top-level in Xcode
-- Follow README.md there to get Xcode roughly set up
-- Go to `Package.swift` and find the dependencies section. Add:
+We are using Uniffi with the latest procedural macros syntax where possible, which also gives us async support. It is important to learn the syntax: https://mozilla.github.io/uniffi-rs/proc_macro/index.html
 
-```swift
-        .package(url: "../xmtp-rust-swift/", branch: "your-local-branch")
-```
+For the most part, any mistakes in the Uniffi interface will manifest as a compile error when running `./gen_kotlin.sh`. Some details are described below so that they are easier to understand.
 
-- Make a local commit in xmtp-rust-swift (must be a commit to get picked up by Swift Package) and reference it in Xcode
-- Note: commits to the same branch will not be picked up by Xcode, you must right-click the Swift package and do "Update Package"
+## Object Lifetimes
+
+Any objects crossing the Uniffi interface boundary must be wrapped in `Arc<>`, so that Uniffi can marshall it back and forth between [raw pointers](https://mozilla.github.io/uniffi-rs/internals/object_references.html#lifetimes) before passing it to the foreign language. The usage of Arc means that we do not need to manually destroy objects on the Rust side, however depending on the target platform, the foreign language may need to automatically or manually [release the pointer back to Rust](https://mozilla.github.io/uniffi-rs/kotlin/lifetimes.html) when done.
+
+## Async and concurrency
+
+We use Tokio as our multi-threaded [async runtime](https://rust-lang.github.io/async-book/08_ecosystem/00_chapter.html). Uniffi can use this runtime on async methods and objects using the annotation `#[uniffi::export(async_runtime = ‘tokio’)]`. Uniffi plumbs up an executor (scheduler) in the foreign language to the Tokio runtime in Rust. More details [here](https://github.com/mozilla/uniffi-rs/blob/734050dbf1493ca92963f29bd3df49bb92bf7fb2/uniffi_core/src/ffi/rustfuture.rs#L11-L18).
+
+Because the foreign language may be multi-threaded, any objects passed to the foreign language must be `Send` and `Sync`, and [no references to `&mut self` are permitted](https://mozilla.github.io/uniffi-rs/udl/interfaces.html#concurrent-access). TODO: interior mutability pattern we should use, sync vs async mutex/rwlock
```

Comment:
This restriction is going to be tricky to work with, but seems like it's necessary with uniffi. I assume this goes all the way down the chain (if I expose a struct that has another struct as a property, that child struct can't have any `&mut self` references either).

Reply:
Or can we maybe work around that by wrapping the child in an Arc or something?